src/v/kafka/security/scram_algorithm.cc

```diff
@@ -47,14 +60,98 @@ namespace kafka {
 // NOLINTNEXTLINE
 #define ALPHA "[a-zA-Z]"
 
-// clang-format off
-static constexpr auto client_first_message_re = ctll::fixed_string{
-    "n,(?:a=(" SASLNAME "))?,(m=" VALUE ",)?n=(" SASLNAME "),"
-    "r=(" PRINTABLE ")(," ALPHA "+=" VALUE ")*"};
+#define CLIENT_FIRST_MESSAGE_RE                                                \
+    "n,(?:a=(" SASLNAME "))?,(m=" VALUE ",)?n=(" SASLNAME "),"                 \
+    "r=(" PRINTABLE ")(," ALPHA "+=" VALUE ")*"
+
+#define CLIENT_FINAL_MESSAGE_RE                                                \
+    "c=(" BASE64 "),r=(" PRINTABLE ")" EXTENSIONS ",p=(" BASE64 ")"
 
+#ifdef USE_CTRE
+static constexpr auto client_first_message_re = ctll::fixed_string{
+  CLIENT_FIRST_MESSAGE_RE};
 static constexpr auto client_final_message_re = ctll::fixed_string{
-    "c=(" BASE64 "),r=(" PRINTABLE ")" EXTENSIONS ",p=(" BASE64 ")"};
-// clang-format on
+  CLIENT_FINAL_MESSAGE_RE};
+#else
+static const char* client_first_message_re = CLIENT_FIRST_MESSAGE_RE;
+static const char* client_final_message_re = CLIENT_FINAL_MESSAGE_RE;
+#endif
+
+struct client_first_match {
+    ss::sstring authzid;
+    ss::sstring username;
+    ss::sstring nonce;
+    ss::sstring extensions;
+};
+
+static inline std::optional<client_first_match>
+parse_client_first(std::string_view message) {
+#ifdef USE_CTRE
+    auto match = ctre::match<client_first_message_re>(message);
+    if (unlikely(!match)) {
+        return std::nullopt;
+    }
+    return client_first_match{
+      .authzid = match.get<1>().to_string(),
+      .username = match.get<3>().to_string(),
+      .nonce = match.get<4>().to_string(),
+      .extensions = match.get<5>().to_string(), // NOLINT
+    };
+#else
+    static const std::regex re(
+      client_first_message_re, std::regex::ECMAScript | std::regex::optimize);
+    std::smatch match;
+    std::string m(message);
+    if (unlikely(!std::regex_match(m, match, re))) {
+        return std::nullopt;
+    }
+    return client_first_match{
+      .authzid = match[1].str(),
+      .username = match[3].str(),
+      .nonce = match[4].str(),
+      .extensions = match[5].str(), // NOLINT
+    };
+#endif
+}
+
+struct client_final_match {
+    bytes channel_binding;
+    ss::sstring nonce;
+    ss::sstring extensions;
+    bytes proof;
+};
+
+static inline std::optional<client_final_match>
+parse_client_final(std::string_view message) {
+#ifdef USE_CTRE
+    auto match = ctre::match<client_final_message_re>(message);
```

Comment:
same idea here. 

would be good to have them forward declared so we can test them in unit tests.